In [1]:
import pandas as pd
import numpy as np
data_path = "./cb_db_csv/"

In [2]:
pd.__version__

'0.23.4'

# cb_objects

In [3]:
df_object = pd.read_csv(data_path + 'cb_objects.csv', encoding = "ISO-8859-1")
df_object = df_object.replace(r'\N', np.nan)
df_object.founded_at = pd.to_datetime(df_object.founded_at)
df_object=df_object[pd.notnull(df_object['founded_at'])]
len(df_object.index)

100441

## Helper Method

In [6]:
TermList = ['angel','seed','a', 'b','c','d','e','f','g']
            # 'convertible','crowd','crowd_equity','debt_round','grant',
            # 'partial','post_ipo_debt','post_ipo_equity','private_equity','secondary_market','unattributed']
def FundMultiHotEncoding(row):
    global TermList 
    termEncoding = np.zeros(len(TermList))
    if not pd.isna(row): 
        if (row in TermList) == True:
            index = TermList.index(row)
            termEncoding[index] = 1
            return termEncoding
        else:
            row = np.nan
    return row

## IPOs

In [245]:
df_ipo = pd.read_csv(data_path + 'cb_ipos.csv', encoding = "ISO-8859-1")
df_ipo = df_ipo.drop(['source_url','created_at','updated_at'],axis=1)
df_ipo = df_ipo.replace(r'\N', np.nan)
## drop object_id = null
## drop companies without id record
df_ipo = df_ipo[df_ipo.object_id.notnull()]
null_columns=df_ipo.columns[df_ipo.isnull().any()]
print(df_ipo[null_columns].isnull().sum())

## make timestamp 
df_ipo.public_at = pd.to_datetime(df_ipo.public_at)
df_ipo['time_since_founded'] = pd.Timedelta('nan')
print(len(df_ipo.index))

valuation_amount           1146
valuation_currency_code       2
raised_amount              1115
raised_currency_code        558
public_at                   595
dtype: int64
1254


- Before: 1254
- After: 426

In [ ]:
def getIPODuration(row):
    df_row = df_object.loc[df_object['id'] == row.object_id]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['public_at']):
            duration = row.public_at - df_row.founded_at
            row['time_since_founded'] = duration.astype('timedelta64[D]').values
    return row
df_ipo = df_ipo.apply(getIPODuration, axis=1)
df_ipo = df_ipo[df_ipo['time_since_founded'].notnull()]
df_ipo = df_ipo.set_index(['object_id']).sort_index()
df_ipo = df_ipo.drop(['id','ipo_id','valuation_amount','valuation_currency_code','raised_amount','raised_currency_code',
                    'public_at','stock_symbol'],axis=1)


In [209]:
df_ipo.to_pickle('./dataframe_cleaned/df_ipo') 

## Acquisition
### ID list and # of acquistions

In [ ]:
df_acq = pd.read_csv(data_path + 'cb_acquisitions.csv', encoding = "ISO-8859-1")
df_acq = df_acq.drop(['source_url','created_at','updated_at'],axis=1)

df_acq = df_acq.replace(r'\N', np.nan)
df_acq = df_acq[df_acq['acquired_object_id'].notnull()]
df_acq = df_acq[df_acq['acquired_at'].notnull()]
null_columns=df_acq.columns[df_acq.isnull().any()]
print(df_acq[null_columns].isnull().sum())

#df_acq['price_currency_code'] = df_acq['price_currency_code'].apply(MultiHotEncoding)
#df_acq['term_code'] = df_acq['term_code'].apply(AcqMultiHotEncoding)
df_acq = pd.concat([df_acq, pd.get_dummies(df_acq['price_currency_code'])], axis=1, sort=False)
df_acq = pd.concat([df_acq, pd.get_dummies(df_acq['term_code'])], axis=1, sort=False)
df_acq

In [211]:
## Acquiring
df_acq.acquired_at = pd.to_datetime(df_acq.acquired_at)
df_acq['acquiring_time_since_founded'] = pd.Timedelta('nan')
def getAcquiringDuration(row):
    df_row = df_object.loc[df_object['id'] == row['acquiring_object_id']]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['acquired_at']):
            duration = row['acquired_at'] - df_row['founded_at']
            row['acquiring_time_since_founded'] = duration.astype('timedelta64[D]').values
    return row
df_acq = df_acq.apply(getAcquiringDuration, axis=1)


In [212]:
## Acquired
df_acq['acquired_time_since_founded'] = pd.Timedelta('nan')
def getAcquiredDuration(row):
    df_row = df_object.loc[df_object['id'] == row['acquired_object_id']]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['acquired_at']):
            duration = row['acquired_at'] - df_row['founded_at']
            row['acquired_time_since_founded'] = duration.astype('timedelta64[D]').values
    return row
df_acq = df_acq.apply(getAcquiredDuration, axis=1)


In [213]:
#print(len())
#df_acq = df_acq[df_acq['acquiring_time_since_founded'].notnull()]
#df_acq = df_acq[df_acq['acquired_time_since_founded'].notnull()]
# save df
df_acq.to_pickle('./dataframe_cleaned/df_acq') 
df_acq

id  acquisition_id acquiring_object_id acquired_object_id  \
0        1               1                c:11               c:10   
1        2               7                c:59               c:72   
2        3               8                c:24              c:132   
3        4               9                c:59              c:155   
4        5              10               c:212              c:215   
5        6              11               c:227              c:228   
6        7              12                 c:5              c:230   
7        8              14               c:318              c:314   
8        9              15                c:20              c:333   
9       10              16               c:342              c:341   
10      11              17                c:24              c:347   
11      12              18               c:349              c:348   
12      13              19               c:355              c:354   
13      14              20               c:377              c:100   
14      15              22               c:351              c:366   
15      16              23               c:307              c:306   
16      17              24               c:212              c:394   
17      18              26               c:499              c:495   
18      19              27                c:29              c:528   
19      20              28                c:29              c:530   
20      21              29               c:541               c:52   
21      22              30               c:541              c:542   
22      23              31               c:568              c:201   
23      24              32               c:570              c:569   
24      25              33                c:20              c:571   
25      26              34               c:317              c:572   
26      27              35               c:573              c:574   
27      28              36               c:583              c:581   
28      29              37               c:584              c:585   
29      30              38               c:590              c:588   
...    ...             ...                 ...                ...   
9532  9533           10496            c:230880           c:285385   
9533  9534           10497             c:50953            c:26342   
9534  9535           10498            c:285396           c:285399   
9535  9536           10499               c:129           c:142814   
9536  9537           10500             c:42365            c:79582   
9537  9538           10501            c:258527           c:285601   
9538  9539           10502            c:258527           c:285603   
9539  9540           10503              c:1434            c:60436   
9540  9541           10505            c:233530           c:285623   
9541  9542           10506            c:285627           c:285630   
9542  9543           10507                c:29           c:169154   
9543  9544           10511            c:205470            c:45837   
9544  9545           10512            c:195465           c:195464   
9545  9546           10513             c:43761            c:43761   
9546  9547           10514            c:285827           c:146509   
9547  9548           10515             c:22784           c:237136   
9548  9549           10516              c:1600            c:71968   
9549  9550           10517            c:285921            c:71996   
9550  9551           10518             c:16591           c:209849   
9551  9552           10519             c:10120           c:285956   
9552  9553           10520               c:672           c:285961   
9553  9554           10521             c:16259           c:283148   
9554  9555           10522            c:285964            c:13710   
9555  9556           10523            c:285968           c:285969   
9556  9557           10524            c:285984           c:285991   
9557  9558           10525            c:267859           c:285995   
9558  

In [ ]:
## Acquiring

df_acquiring = df_acq.set_index(['acquiring_object_id','acquired_object_id']).sort_index()
df_acquiring.sort_values(by=['acquiring_time_since_founded'],inplace=True)
df_acquiring.drop(['id','acquisition_id','acquired_at','acquired_time_since_founded'],axis=1,inplace=True)
df_acquiring = df_acquiring[df_acquiring['acquiring_time_since_founded'].notnull()]
df_acquiring = df_acquiring[df_acquiring['acquiring_time_since_founded'].gt(0)]


In [222]:
df_acquiring = df_acquiring[df_acquiring['price_amount'].notnull()]

df_acquiring = df_acquiring.drop(['price_currency_code','term_code'],axis=1)

In [223]:
df_acquiring.to_pickle('./dataframe_cleaned/df_acquiring') 

In [224]:
df_acquiring

price_amount  AUD  CAD  EUR  GBP  JPY  \
acquiring_object_id acquired_object_id                                         
c:60326             c:146208                1000000    0    0    0    0    0   
c:15461             c:15460                   70000    0    0    0    0    0   
c:192899            c:193355                 330000    0    0    0    0    0   
c:138299            c:141275                  15000    0    0    0    0    0   
c:5457              c:5456                     5000    0    0    0    0    0   
c:195859            c:195858                 100000    0    0    0    0    0   
c:221698            c:190514                   5000    0    0    0    0    0   
c:171432            c:44839                    3000    0    0    0    0    0   
c:9212              c:9211                    50000    0    0    1    0    0   
c:185857            c:182355                   2500    0    0    0    0    0   
c:184438            c:184561                    200    0    0    0    0    0   
c:73652             c:85064                30000000    0    0    0    0    0   
c:475               c:5721                 10500000    0    0    0    0    0   
c:26413             c:29725                 1800000    0    0    0    0    0   
c:41425             c:42612                   50000    0    0    0    0    0   
c:35053             c:15882                   10000    0    0    0    0    0   
c:266298            c:266297                      1    0    0    0    0    0   
c:236444            c:236433               24000000    0    0    0    0    0   
c:215031            c:215029                6600000    0    0    0    0    0   
c:199561            c:13199                   30000    0    0    1    0    0   
c:22119             c:22116                   12000    0    0    0    0    0   
c:215031            c:215030                7590000    0    0    0    0    0   
c:25574             c:156253                  62000    0    0    0    0    0   
c:42080             c:140823                5200000    0    0    0    0    0   
c:18366             c:18365                 3800000    0    0    0    0    0   
c:242649            c:242649                   1858    0    0    0    0    0   
c:261276            c:261265              185000000    0    0    0    0    0   
c:475               c:13918                  836000    0    0    0    0    0   
c:226873            c:226871                  75000    0    0    0    0    0   
                    c:226872                 279000    0    0    0    0    0   
...                                             ...  ...  ...  ...  ...  ...   
c:258527            c:285601              260000000    0    0    0    0    0   
                    c:285603               28000000    0    0    0    0    0   
c:26486             c:26487               124000000    0    1    0    0    0   
                    c:478                   3800000    0    0    0    0    0   
c:13779             c:268949             1200000000    0    0    0    0    0   
c:27295             c:282030               48600000    0    0    0    1    0   
c:21683             c:4002                374000000    0    0    0    0    0   
c:7389              c:28626                 8000000    0    0    0    0    0   
c:83647             c:28502               266000000    0    0    0    0    0   
c:26486             c:3445                 27700000    0    0    0    0    0   
                    c:3525                 27000000    0    0    0    0    0   
c:7060              c:50055                54000000    0    0    0    0    0   
c:31279             c:31278                98300000    0    0    0    0    0   
c:1246              c:596                  45000000    0    0    0    0    0   
c:59312             c:228074              195000000    0    0    0    0    0   
c:31357             c:32180               175000000    0    0    0    0    0   
c:259799            c:183726               30400000    0    0    0    0    0   
c:7060              c:25775            

Before kicking away the negative durations 
- 9532 rows × 4 columns 

After kicking away the negative durations 
- 6866 rows × 4 columns

After dropping entries with missing values price amount
- 1875 rows × 4 columns

## Acquired

Before kicking away the negative durations 
- 9532 rows × 4 columns

After kicking away the negative durations 
- 4821 rows × 4 columns


In [225]:
df_acquired = df_acq.set_index(['acquired_object_id']).sort_index()
df_acquired.sort_values(by=['acquired_time_since_founded'],inplace=True)
df_acquired.drop(['id','acquisition_id','acquiring_object_id','acquired_at','acquiring_time_since_founded','term_code','price_amount','price_currency_code'],axis=1,inplace=True)
df_acquired = df_acquired[df_acquired['acquired_time_since_founded'].gt(0)]

df_acquired.to_pickle('./dataframe_cleaned/df_acquired') 

In [226]:
df_acquired

AUD  CAD  EUR  GBP  JPY  SEK  USD  cash  cash_and_stock  \
acquired_object_id                                                            
c:22103               0    0    0    0    0    0    1     1               0   
c:54651               0    0    0    0    0    0    1     0               0   
c:2315                0    0    0    0    0    0    1     0               0   
c:329                 0    0    0    0    0    0    1     0               0   
c:159552              0    0    0    0    0    0    1     0               0   
c:54405               0    0    0    0    0    0    1     0               0   
c:8565                0    0    0    0    0    0    1     0               0   
c:86                  0    0    0    0    0    0    1     0               0   
c:45891               0    0    0    0    0    0    1     1               0   
c:193193              0    0    0    0    0    0    1     0               0   
c:681                 0    0    0    0    0    0    1     1               0   
c:192352              0    0    0    0    0    0    1     0               0   
c:149002              0    0    0    0    0    0    1     0               0   
c:47111               0    0    0    0    0    0    1     1               0   
c:195841              0    0    0    0    0    0    1     0               0   
c:3306                0    0    0    0    0    0    1     1               0   
c:2370                0    0    0    0    0    0    1     0               1   
c:213362              0    0    0    0    0    0    1     0               0   
c:27053               0    0    0    0    0    0    1     0               0   
c:1229                0    0    0    0    0    0    1     0               1   
c:156584              0    0    0    0    0    0    1     0               0   
c:175581              0    0    0    0    0    0    1     0               0   
c:212610              0    0    0    0    0    0    1     0               0   
c:183206              0    0    0    0    0    0    1     0               0   
c:211517              0    0    0    0    0    0    1     0               0   
c:187849              0    0    0    0    0    0    1     0               0   
c:3138                0    0    0    0    0    0    1     0               0   
c:182466              0    0    0    0    0    0    1     0               0   
c:8652                0    0    0    0    0    0    1     1               0   
c:245790              0    0    0    0    0    0    1     0               0   
...                 ...  ...  ...  ...  ...  ...  ...   ...             ...   
c:30506               0    0    0    0    0    0    1     0               0   
c:269416              0    0    0    0    0    0    1     0               0   
c:38641               0    0    0    0    0    0    1     0               0   
c:242705              0    0    0    0    0    0    1     0               0   
c:216499              0    0    0    0    0    0    1     0               0   
c:25631               0    0    0    0    0    0    1     0               0   
c:32156               0    0    0    0    0    0    1     0               0   
c:29313               0    0    0    0    0    0    1     1               0   
c:267421              0    0    0    0    0    0    1     0               0   
c:5296                0    0    0    0    0    0    1     0               0   
c:34009               0    0    0    0    0    0    1     0               0   
c:2332                0    0    0    0    0    0    1     0               1   
c:51460               0    0    0    0    0    0    1     0               0   
c:81111               0    0    0    0    0    0    1     1               0   
c:86083               0    0    0    0    0    0    1     1               0   
c:10232               0    0    0    0    0    0    1     0               0   
c:270108              0    0    0    0    0    0    1     1               0   
c:220985              0    0    0    0    0    0    1     0           

## Funding
### Todo: Replace entry missing funded_at and raised_amount_usd with -1

Non dropping raied amount and funded at 
- 31939 companies

After
- 27985 companies

After funding round code
- 18499 companies

After funding round code
- 18499 companies

In [228]:
df_funding= pd.read_csv(data_path + 'cb_funding_rounds.csv', encoding = "ISO-8859-1")
df_funding = df_funding.drop(['source_url','created_by','created_at','updated_at'],axis=1)
df_funding = df_funding.replace(r'\N', np.nan)

df_funding = df_funding[pd.notnull(df_funding['raised_amount_usd'])]
df_funding = df_funding[pd.notnull(df_funding['funded_at'])]

null_columns = df_funding.columns[df_funding.isnull().any()]
print(df_funding[null_columns].isnull().sum())

27985
pre_money_valuation_usd     46777
pre_money_valuation         46777
pre_money_currency_code     22090
post_money_valuation_usd    45650
post_money_valuation        45649
post_money_currency_code    19299
dtype: int64


In [ ]:
#drop those funding rounds out of concern
df_funding = df_funding[df_funding['funding_round_code'].isin(TermList)]
df_funding = pd.concat([df_funding, pd.get_dummies(df_funding['funding_round_code'])], axis=1, sort=False)

In [243]:
## make timestamp 
df_funding.funded_at = pd.to_datetime(df_funding.funded_at)
df_funding['time_since_founded'] = pd.Timedelta('nan')

In [244]:
def getDuration(row):
    df_row = df_object.loc[df_object['id'] == row.object_id]
    if not df_row.empty:
        if pd.notnull(df_row['founded_at'].values) and pd.notnull(row['funded_at']):
            duration = row.funded_at - df_row.founded_at
            row['time_since_founded'] = duration.astype('timedelta64[D]').values

    return row

df_funding = df_funding.apply(getDuration,axis=1)

## Funding Data
- Before kicking away NAN and negative durations: 31939 companies
- After: 14691 companies 

In [ ]:
#only positive date
df_funding = df_funding[df_funding['time_since_founded'].gt(-0.1)]
df_funding['time_since_founded'] = df_funding['time_since_founded'].astype(int)

In [249]:
df_funding = df_funding.set_index(['object_id','time_since_founded'])
df_funding.drop(['id','funding_round_id','funded_at','raised_amount','raised_currency_code','pre_money_valuation_usd','pre_money_valuation',
                 'pre_money_currency_code','post_money_valuation_usd','post_money_valuation',
                 'post_money_currency_code'],axis=1,inplace=True)

In [250]:
df_funding

funding_round_type funding_round_code  \
object_id time_since_founded                                         
c:4       781                          series-b                  b   
c:5       213                             angel              angel   
          455                          series-a                  a   
          790                          series-b                  b   
c:7299    1065                         series-b                  b   
c:9       214                          series-a                  a   
          273                          series-b                  b   
c:13      1399                         series-a               seed   
c:14      1581                         series-a                  a   
c:15      0                               angel              angel   
c:16      516                          series-b                  b   
          516                          series-a                  a   
c:15      151                          series-a                  a   
c:22      151                          series-a                  a   
c:23      220                          series-a                  a   
c:27      226                          series-b                  b   
c:30      243                          series-a                  a   
c:31      1461                        series-c+                  c   
          823                          series-b                  b   
c:34      851                          series-a                  a   
          1127                         series-b                  b   
c:27      501                         series-c+                  c   
          1537                        series-c+                  d   
c:38      212                          series-a                  a   
c:40      821                          series-b                  b   
          1308                        series-c+                  c   
c:43      182                             angel               seed   
c:44      395                         series-c+                  c   
          91                           series-a                  a   
          212                          series-b                  b   
...                                         ...                ...   
c:285830  30                           series-a               seed   
c:179637  955                         series-c+                  c   
c:184680  311                          series-a                  a   
c:167539  1787                         series-a                  a   
          2412                         series-b                  b   
c:73717   1136                         series-a                  a   
c:36444   6249                         series-a                  a   
c:275974  1075                         series-a                  a   
c:151059  1044                            angel               seed   
c:285592  83                              angel               seed   
c:38811   3006                        series-c+                  c   
c:203158  350                             angel               seed   
c:171632  756                             angel               seed   
          638                             angel               seed   
          453                             angel               seed   
c:23227   1243                        series-c+                  c   
          882                          series-a                  a   
c:284350  344                             angel               seed   
c:236066  2171                         series-a                  a   
c:27385   1276                         series-b                  b   
c:38711   464                          series-b                  b   
c:181145  559                          series-b                  b   
c:161321  1257                         series-a                  a   
          1093                            angel               seed   
c:17321   2654                         

In [251]:
len(df_funding.groupby('object_id').groups)

14691

In [252]:
df_funding.to_pickle('./dataframe_cleaned/df_funding')